contracts/token/ERC1155/ERC1155MetadataURICatchAll.sol

```diff
@@ -0,0 +1,46 @@
+pragma solidity ^0.6.0;
+
+import "./ERC1155.sol";
+import "./IERC1155MetadataURI.sol";
+import "../../introspection/ERC165.sol";
+
+contract ERC1155MetadataURICatchAll is ERC165, ERC1155, IERC1155MetadataURI {
+    // Catch-all URI with placeholders, e.g. https://example.com/{locale}/{id}.json
+    string private _uri;
+
+     /*
+     *     bytes4(keccak256('uri(uint256)')) == 0x0e89341c
+     */
+    bytes4 private constant _INTERFACE_ID_ERC1155_METADATA_URI = 0x0e89341c;
+
+    /**
+     * @dev Constructor function
+     */
+    constructor (string memory uri) public {
+        _setURI(uri);
+
+        // register the supported interfaces to conform to ERC1155 via ERC165
+        _registerInterface(_INTERFACE_ID_ERC1155_METADATA_URI);
+    }
+
+    /**
+     * @notice A distinct Uniform Resource Identifier (URI) for a given token.
+     * @dev URIs are defined in RFC 3986.
+     * The URI MUST point to a JSON file that conforms to the "ERC-1155 Metadata URI JSON Schema".
+     * param id uint256 ID of the token to query (ignored in this particular implementation,
+     * as an {id} parameter in the string is expected)
+     * @return URI string
+    */
+    function uri(uint256 /*id*/) external view override returns (string memory) {
+        return _uri;
+    }
+
+    /**
+     * @dev Internal function to set a new URI
+     * @param newuri New URI to be set
+     */
+    function _setURI(string memory newuri) internal virtual {
+        _uri = newuri;
+        emit URI(_uri, 0);
```

Comment:
This seems sensible, but I couldn't find anything in the specification about the event's `id` argument in cases such as this. Given that an `id` value of 0 is valid, perhaps we shouldn't emit the event at all? The spec seems to allow this:

>Changes to the URI MUST emit the URI event if the change can be expressed with an event (i.e. it isn’t dynamic/programmatic).
>
>    An implementation MAY emit the URI event during a mint operation but it is NOT mandatory. An observer MAY fetch the metadata uri at mint time from the uri function if it was not emitted.
>
>The uri function SHOULD be used to retrieve values if no event was emitted.

Reply:
Oh, you mean "it isn't dynamic" says that the `{id}` pattern doesn't need it? I interpreted it as it can only be omitted if e.g. the URI is built by a function that depends on, say, other fields and the actually returned URI changes dynamically - but in our case, it can be expressed by an event, and the actually returned URI doesn't change dynamically, it changes only with that call. I have wondered about the ID in that event as well though, unfortunately, the spec doesn't define this case. Maybe we should ask there.

Reply:
No, I agree with your interpretation and believe this case might be missing from the spec. Emitting an event feels wrong though because there is no invalid values for the `id` field we can use to signal what is going on. Asking the spec authors sounds like the way to go.

Reply:
At any rate, given the spec allows for no events to be emitted (like in the case of dynamic URIs), I think it makes sense to not emit anything here.

Reply:
> At any rate, given the spec allows for no events to be emitted (like in the case of dynamic URIs), I think it makes sense to not emit anything here.

We definitely interpret the spec differently there, as I read it as non-optional, with the only exception being where it is not static and not possible to emit the event - and it is static in our case and very much possible to emit it. But if you as the reviewer(s) prefer that, I'll remove it even if I disagree. ;-)

Reply:
We've also discussed this bit with @frangio and agree that we should not emit any events. The spec clearly allows for situations where no events have been emitted, as seen in these statements:

>The uri function SHOULD be used to retrieve values if no event was emitted.
>The uri function MUST return the same value as the latest event for an _id if it was emitted.